geoadmin/src/main/java/org/activityinfo/geoadmin/match/HungarianAlgorithm.java

```diff
@@ -0,0 +1,349 @@
+        reduce();
+        computeInitialFeasibleSolution();
+        greedyMatch();
+
+        int w = fetchUnmatchedTarget();
+        while (w < dim) {
+            initializePhase(w);
+            executePhase();
+            w = fetchUnmatchedTarget();
+        }
+        int[] result = Arrays.copyOf(matchSourceByTarget, rows);
+        for (w = 0; w < result.length; w++) {
+            if (result[w] >= cols) {
+                result[w] = -1;
+            }
+        }
+        return result;
+    }
+
+    /**
+     * Execute a single phase of the algorithm. A phase of the Hungarian algorithm
+     * consists of building a set of committed targets and a set of committed sources
+     * from a root unmatched target by following alternating unmatched/matched
+     * zero-slack edges. If an unmatched source is encountered, then an augmenting
+     * path has been found and the matching is grown. If the connected zero-slack
+     * edges have been exhausted, the labels of committed targets are increased by
+     * the minimum slack among committed targets and non-committed sources to create
+     * more zero-slack edges (the labels of committed sources are simultaneously
+     * decreased by the same amount in order to maintain a feasible labeling).
+     * <p>
+     *
+     * The runtime of a single phase of the algorithm is O(n^2), where n is the
+     * dimension of the internal square cost matrix, since each edge is visited at
+     * most once and since increasing the labeling is accomplished in time O(n) by
+     * maintaining the minimum slack values among non-committed sources. When a phase
+     * completes, the matching will have increased in size.
+     */
+    protected void executePhase() {
+        while (true) {
+            int minSlackTarget = -1, minSlackSource = -1;
+            double minSlackValue = Double.POSITIVE_INFINITY;
+            for (int j = 0; j < dim; j++) {
+                if (parentTargetByCommittedSource[j] == -1) {
+                    if (minSlackValueBySource[j] < minSlackValue) {
+                        minSlackValue = minSlackValueBySource[j];
+                        minSlackTarget = minSlackTargetBySource[j];
+                        minSlackSource = j;
+                    }
+                }
+            }
+            if (minSlackValue > 0) {
+                updateLabeling(minSlackValue);
+            }
+            parentTargetByCommittedSource[minSlackSource] = minSlackTarget;
+            if (matchTargetBySource[minSlackSource] == -1) {
+        /*
+         * An augmenting path has been found.
+         */
+                int committedSource = minSlackSource;
+                int parentTarget = parentTargetByCommittedSource[committedSource];
+                while (true) {
+                    int temp = matchSourceByTarget[parentTarget];
+                    match(parentTarget, committedSource);
+                    committedSource = temp;
+                    if (committedSource == -1) {
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Refactor this code to not nest more than 3 if/for/while/switch/try statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS134)